### PR TITLE
Feature/4030 new warnothers screen

### PR DESCRIPTION
**Pull request by whiskey**
## Description
This PR implements the new 'warn others' view and adds legal texts. 

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4030

## Screenshot
Simulator Screen Shot - iPhone 12 mini - 2020-12-01 at 13 09 25
Simulator Screen Shot - iPhone 12 mini - 2020-12-01 at 13 09 29
Simulator Screen Shot - iPhone 12 mini - 2020-12-01 at 13 09 33

### CODE REVIEW

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewModel.swift [hunk 1]
```diff
@@ -123,15 +125,23 @@ struct ExposureSubmissionQRInfoViewModel {
 }
 
 extension DynamicCell {
-	
+
+	/// A `DynamicAcknowledgementCell` to display a list of acknowledgements the user is informed about.
+	/// - Parameters:
+	///   - title: The title/header for the legal foo.
+	///   - description: Optional description text.
+	///   - bulletPoints: A list of strings to be prefixed with bullet points.
```
Comment (marcussc) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewModel.swift:133]: 👍

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewModel.swift [hunk 2]
```diff
@@ -6,102 +6,128 @@ import Foundation
...
 
 struct ExposureSubmissionWarnOthersViewModel {
-	
-	// MARK: - Init
 
-	init(
-		supportedCountries: [Country]
-	) {
-		self.supportedCountries = supportedCountries.sorted { $0.localizedName.localizedCompare($1.localizedName) == .orderedAscending }
+	private let countries: [Country]
```
Comment (marcussc) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewModel.swift:10]: Please move it to the private section

**Review by marcussc — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewModel.swift [hunk 3]
```diff
@@ -6,102 +6,128 @@ import Foundation
...
+	private let acknowledgementString: NSAttributedString = {
+		let boldText = AppStrings.ExposureSubmissionWarnOthers.acknowledgement_1_1
+		let normalText = AppStrings.ExposureSubmissionWarnOthers.acknowledgement_1_2
+		let string = NSMutableAttributedString(string: "\(boldText) \(normalText)")
```
Comment (tcfos) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewModel.swift:14]: Is 'string' a good name?

**Review by tcfos — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewModel.swift (hunk 3 above)
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewModel.swift:14]: Within that scope - sure!

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewController.swift [hunk 4]
```diff
@@ -43,7 +43,7 @@ class ExposureSubmissionQRInfoViewController: DynamicTableViewController {
 	// MARK: - Internal
 
 	enum ReuseIdentifiers: String, TableViewCellReuseIdentifiers {
-		case acknowledgement = "DynamicAcknowledgementCell"
+		case legal = "DynamicAcknowledgementCell"
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewController.swift:46]:
```suggestion
		case legal = "DynamicLegalCell"
```
I was wondering why the cell name identifiers stays unchanged

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewController.swift [hunk 5]
```diff
@@ -70,7 +70,7 @@ class ExposureSubmissionQRInfoViewController: DynamicTableViewController {
 
 		tableView.register(
 			UINib(nibName: String(describing: DynamicAcknowledgementCell.self), bundle: nil),
-			forCellReuseIdentifier: ReuseIdentifiers.acknowledgement.rawValue
+			forCellReuseIdentifier: ReuseIdentifiers.legal.rawValue
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewController.swift:73]: If think if the case was changed to .legal the Cell class it self should be called something like 'DynamicLegalCell' ?

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift [hunk 6]
```diff
@@ -78,6 +89,19 @@ extension ExposureSubmissionWarnOthersViewController {
...
+extension ExposureSubmissionWarnOthersViewController: ENANavigationControllerWithFooterChild {
+
+	func navigationController(_ navigationController: ENANavigationControllerWithFooter, didTapPrimaryButton button: UIButton) {
+		onPrimaryButtonTap { [weak self] isLoading in
+			DispatchQueue.main.async {
+				self?.navigationFooterItem?.isPrimaryButtonLoading = isLoading
+				self?.navigationFooterItem?.isPrimaryButtonEnabled = !isLoading
+			}
+		}
+	}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift:102]: please keep the code inside the Mark section as it was before

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift [hunk 7]
```diff
@@ -31,43 +31,54 @@ class ExposureSubmissionWarnOthersViewController: DynamicTableViewController, EN
...
+		navigationFooterItem
+	}
 
-	func navigationController(_ navigationController: ENANavigationControllerWithFooter, didTapPrimaryButton button: UIButton) {
-		onPrimaryButtonTap { [weak self] isLoading in
-			DispatchQueue.main.async {
-				self?.navigationFooterItem?.isPrimaryButtonLoading = isLoading
-				self?.navigationFooterItem?.isPrimaryButtonEnabled = !isLoading
-			}
-		}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift:44]: Why remove this here?

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift (hunk 7 above)
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift:44]: As in `ExposureSubmissionQRInfoViewController`: code style. By moving everything protocol-related into a dedicated extension you can move out or refactor protocol implementations way faster and easier than looking up all functions in the current MARK-nonsense.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift (hunk 6 above)
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift:102]: [see above](https://github.com/corona-warn-app/cwa-app-ios/pull/1609/files/246bf7f83919fb6cda893514be64ce60c5d974b7#r533956426)

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewController.swift (hunk 5 above)
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionQRInfoViewController.swift:73]: The cell is renamed but in this context I find 'acknowledgement' more appropriate.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift (hunk 7 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift:44]: But reading the code is more complex if you open up the file you can't immediately see which protocols are implemented. You have to search all over the project for possible extensions. I also don't see why refactoring should be faster that way? Copying from an extensions section is basically the same effort as copying from a Mark section. 
Also this file was change just before weeks to the MARK-nonsense (how you call it) and now it gets changed back again. As agreed in the dev camp we decided to use MARKs. If you like to work in a different style, make a request and we can vote on what the majority wants.

Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionWarnOthersViewController.swift:44]: > But reading the code is more complex if you open up the file you can't immediately see which protocols are implemented. You have to search all over the project for possible extensions. 

You don't read the code like a book, don't you? If you check for references, symbols or even text-matches you'll end at the correct place, even if further down in the code.

> I also don't see why refactoring should be faster that way? Copying from an extensions section is basically the same effort as copying from a Mark section.

If I need some private helper functions I have to clutter them into other sections not related to the protocol implementation.

The only downside might come from very large protocols but monstrosities like `ExposureSubmissionService` should be split up in future. We can structure code like in `AppConfigurationProviding` and `AppConfigurationFetching` and even group it like in `Store`. 

Request for discussion is open.

**Review by KaiTeuber — APPROVED** [review]

**Pull request merged**